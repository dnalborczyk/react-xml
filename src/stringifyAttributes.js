import { isPrimitive } from './utils.js'

const { entries } = Object

export default function stringifyAttributes(obj) {
  return entries(obj)
    .filter(([, value]) => value != null)
    .map(([key, value]) => {
      if (isPrimitive(value)) {
        return `${key}="${value}"`
      }

      throw new Error(
        `Unsupported type: ${typeof value}. Only bigint, boolean, number and string are supported.`,
      )
    })
    .join(' ')
}
