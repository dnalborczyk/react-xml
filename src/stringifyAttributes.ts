import isPrimitive from './isPrimitive'

const { entries } = Object

export default function stringifyAttributes(obj: any): string {
  return entries(obj)
    .filter(([, value]) => value != null)
    .map(([key, value]) => {
      if (isPrimitive(value)) {
        return `${key}="${String(value)}"`
      }

      throw new Error(
        `Unsupported type: ${typeof value}. Only bigint, boolean, number and string are supported.`,
      )
    })
    .join(' ')
}
