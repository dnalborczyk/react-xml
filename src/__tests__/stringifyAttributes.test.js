import stringifyAttributes from '../stringifyAttributes.js'

describe('stringifyAttributes tests', () => {
  describe('should work with supported property types', () => {
    test.each`
      param                       | expected               | type
      ${{}}                       | ${''}                  | ${'empty Object'}
      ${{ foo: 1 }}               | ${'foo="1"'}           | ${'Object with number property'}
      ${{ foo: 'bar' }}           | ${'foo="bar"'}         | ${'Object with string property'}
      ${{ foo: '' }}              | ${'foo=""'}            | ${'Object with empty string property'}
      ${{ foo: null }}            | ${''}                  | ${'Object with null property'}
      ${{ foo: undefined }}       | ${''}                  | ${'Object with undefined property'}
      ${{ foo: '0', bar: 'foo' }} | ${'foo="0" bar="foo"'} | ${'Object with multiple properties'}
    `('should return $expected when $type is passed', ({ param, expected }) => {
      const result = stringifyAttributes(param)
      expect(result).toEqual(expected)
    })
  })

  describe('should throw with unsupported property types', () => {
    test.each`
      param                         | type
      ${{ foo: [] }}                | ${'Array'}
      ${{ foo: {} }}                | ${'Object'}
      ${{ foo: Promise.resolve() }} | ${'Promise'}
      ${{ foo: Symbol('') }}        | ${'Symbol'}
    `('should throw exception with property type $type', ({ param }) => {
      expect(() => stringifyAttributes(param)).toThrowErrorMatchingSnapshot()
    })
  })
})
