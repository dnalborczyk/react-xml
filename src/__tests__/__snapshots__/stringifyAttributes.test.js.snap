// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stringifyAttributes should throw with unsupported property types should throw exception with property type Array 1`] = `"Unsupported type: object. Only bigint, boolean, number and string are supported."`;

exports[`stringifyAttributes should throw with unsupported property types should throw exception with property type Object 1`] = `"Unsupported type: object. Only bigint, boolean, number and string are supported."`;

exports[`stringifyAttributes should throw with unsupported property types should throw exception with property type Promise 1`] = `"Unsupported type: object. Only bigint, boolean, number and string are supported."`;

exports[`stringifyAttributes should throw with unsupported property types should throw exception with property type Symbol 1`] = `"Unsupported type: symbol. Only bigint, boolean, number and string are supported."`;
